hathor/indexes/utxo_index.py

```diff
@@ -0,0 +1,326 @@
+        - mark transaction as added
+        - inputs are removed from the index
+        - outputs are added to the index
+        """
+        tx_meta = tx.get_metadata()
+        if tx_meta.voided_by:
+            self.log.debug('skip add utxos of voided tx', tx=tx.hash_hex)
+            return
+        self.log.debug('add utxos of', tx=tx.hash_hex)
+        heightlock: Optional[int]
+        timelock: Optional[int]
+        assert tx.hash is not None
+        # ignore if already in the index
+        if self._is_in_index(tx.hash):
+            self.log.warn('already in UTXO index, skip add', tx=tx.hash_hex)
+            return
+        self._add_tx(tx.hash)
+        tx_meta = tx.get_metadata()
+        # inputs are spent, remove them from the index
+        for tx_input in tx.inputs:
+            spent_tx = tx.get_spent_tx(tx_input)
+            spent_tx_meta = spent_tx.get_metadata()
+            spent_tx_output = spent_tx.outputs[tx_input.index]
+            address_script = parse_address_script(spent_tx_output.script)
+            if not address_script:
+                # XXX: completely ignore scripts that aren't "addressable"
+                continue
+            if spent_tx.is_block:
+                # XXX: timelock forced to None when there is a heightlock
+                timelock = None
+                heightlock = spent_tx_meta.height + settings.REWARD_SPEND_MIN_BLOCKS
+            else:
+                timelock = address_script.get_timelock()
+                heightlock = None
+            assert timelock is None or heightlock is None
+            self._remove_utxo(UtxoIndexItem(
+                token_uid=spent_tx.get_token_uid(spent_tx_output.get_token_index()),
+                tx_id=tx_input.tx_id,
+                index=tx_input.index,
+                address=address_script.address,
+                amount=spent_tx_output.value,
+                timelock=timelock,
+                heightlock=heightlock,
+            ))
+        # pre-calculate heightlock, if any, because if there is one it's the same for all outputs
+        if tx.is_block:
+            heightlock = tx_meta.height + settings.REWARD_SPEND_MIN_BLOCKS
+        else:
+            heightlock = None
+        # add new outputs to the index
+        for index, tx_output in enumerate(tx.outputs):
+            address_script = parse_address_script(tx_output.script)
+            if not address_script:
+                # XXX: completely ignore scripts that aren't "addressable"
+                continue
+            if heightlock is None:
+                timelock = address_script.get_timelock()
+            else:
+                # XXX: timelock forced to None when there is a heightlock
+                timelock = None
+            # XXX: that is, at least one of them must but None
+            assert timelock is None or heightlock is None
+            self._add_utxo(UtxoIndexItem(
```

Comment:
Makes sense. Also, what about data-only outputs? Is there a more general indicator? Maybe checking if the amount is 0?

Reply:
For data outputs the amount is actually 1. We should ignore data script output in this index because they can't be spent and you are already doing it in the address check.

If you prefer being more explicit about it, you can do `DataScript.parse_script(output_script)`. We have this class in the python hathorlib. But you are already skipping them, so I feel it's fine

Reply:
It's now explicitly ignoring authority outputs, but I left data outputs to be implicitly being ignored (since the script parser won't be able to get an address out of it).